File: src/lib/linkedin-jobs.ts
/**
 * LinkedIn active job board (guest API) — Israel, posted in the last 7 days.
 * Used by the in-app pipeline and aligned with the twice-daily Actions scan.
 */

import { isIsraelLocation } from "@/lib/israel";
import { extractEmails } from "@/lib/apply-email";
import { extractExternalApplyUrl } from "@/lib/web-apply";
import {
  isRealLinkedInJobId,
  linkedInJobViewUrl,
  normalizeLinkedInOpenUrl,
} from "@/lib/linkedin-url";

export type LinkedInJobRow = {
  source: string;
  external_id: string;
  title: string;
  company: string | null;
  location: string | null;
  url: string;
  description: string;
  apply_email: string | null;
  post_kind: "job";
  channel: "linkedin";
  is_social: boolean;
  scraped_at: string;
  posted_at: string | null;
};

const UA =
  "Mozilla/5.0 (compatible; ai-agent-job-scanner/1.0; +https://github.com/asaf2310-boop/ai-agent)";

/** Past week filter on LinkedIn guest search */
const PAST_WEEK = "r604800";
/** Israel geoId used by LinkedIn public search */
const ISRAEL_GEO = "101620260";

const DEFAULT_SEARCHES = [
  "\"AI product manager\" OR \"AI product owner\" OR \"מנהל מוצר AI\" OR \"LLM product\"",
  "\"product manager\" AND (AI OR LLM OR \"Gen AI\" OR \"בינה מלאכותית\")",
  "\"product owner\" AND (AI OR LLM OR \"machine learning\")",
  "\"product analyst\" OR \"אנליסט מוצר\"",
];

const RELEVANT =
  /ai|llm|machine learning|data scientist|data analyst|product manager|product owner|product analyst|finance|fintech|fp&a|controller|marketing|customer success|מנהל מוצר|בעל מוצר|פיננס|בינה|שיווק|הצלחת לקוחות|אנליסט מוצר/i;

function decodeHtml(s: string): string {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&nbsp;/g, " ")
    .trim();
}

function stripTags(html: string): string {
  return decodeHtml(html.replace(/<[^>]+>/g, " ").replace(/\s+/g, " ")).trim();
}

function parseListHtml(html: string): Array<{
  id: string;
  title: string;
  company: string;
  location: string;
  url: string;
  postedAt: string | null;
}> {
  const jobs: Array<{
    id: string;
    title: string;
    company: string;
    location: string;
    url: string;
    postedAt: string | null;
  }> = [];

  const cards = html.split(/data-entity-urn="urn:li:jobPosting:/).slice(1);
  for (const card of cards) {
    const id = card.match(/^(\d+)/)?.[1];
    if (!id) continue;
    const title = stripTags(
      card.match(/base-search-card__title[^>]*>([\s\S]*?)<\/h3>/i)?.[1] || "",
    );
    const company = stripTags(
      card.match(/base-search-card__subtitle[\s\S]*?<a[^>]*>([\s\S]*?)<\/a>/i)?.[1] ||
        "",
    );
    const location = stripTags(
      card.match(/job-search-card__location[^>]*>([\s\S]*?)<\/span>/i)?.[1] || "",
    );
    const href =
      card.match(
        /base-card__full-link[^>]*href="([^"]+)"/i,
      )?.[1] || "";
    const url = normalizeLinkedInOpenUrl(
      href ? decodeHtml(href).split("?")[0] : null,
      { externalId: id, title },
    );
    const postedAt =
      card.match(/<time[^>]*datetime="([^"]+)"/i)?.[1] || null;

    if (!title) continue;
    jobs.push({ id, title, company, location, url, postedAt });
  }
  return jobs;
}

async function fetchSearchPage(
  keywords: string,
  start: number,
): Promise<string> {
  const params = new URLSearchParams({
    keywords,
    location: "Israel",
    geoId: ISRAEL_GEO,
    f_TPR: PAST_WEEK,
    start: String(start),
    count: "25",
  });
  const res = await fetch(
    `https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search?${params}`,
    {
      headers: {
        "User-Agent": UA,
        Accept: "text/html",
      },
      next: { revalidate: 0 },
    },
  );
  if (!res.ok) return "";
  return res.text();
}

async function fetchJobDescription(jobId: string): Promise<string> {
  try {
    const res = await fetch(
      `https://www.linkedin.com/jobs-guest/jobs/api/jobPosting/${jobId}`,
      {
        headers: { "User-Agent": UA, Accept: "text/html" },
        signal: AbortSignal.timeout(8000),
        next: { revalidate: 0 },
      },
    );
    if (!res.ok) return "";
    const html = await res.text();
    // Prefer description section if present
    const descBlock =
      html.match(
        /show-more-less-html__markup[\s\S]*?>([\s\S]*?)<\/div>/i,
      )?.[1] || html;
    return stripTags(descBlock).slice(0, 4000);
  } catch {
    return "";
  }
}

function withinLastWeek(postedAt: string | null): boolean {
  if (!postedAt) return true; // LinkedIn already filtered f_TPR=r604800
  const t = Date.parse(postedAt);
  if (Number.isNaN(t)) return true;
  const weekMs = 7 * 24 * 60 * 60 * 1000;
  return Date.now() - t <= weekMs + 12 * 60 * 60 * 1000; // small skew buffer
}

/** Fetch Israel LinkedIn jobs from the past week across relevant domains. */
export async function fetchLinkedInIsraelJobs(options?: {
  maxJobs?: number;
  enrichDescriptions?: number;
  searches?: string[];
}): Promise<LinkedInJobRow[]> {
  const maxJobs = options?.maxJobs ?? 80;
  const enrichN = options?.enrichDescriptions ?? 25;
  const searches =
    options?.searches?.length ? options.searches : DEFAULT_SEARCHES;
  const now = new Date().toISOString();
  const byId = new Map<string, LinkedInJobRow>();

  for (const keywords of searches) {
    for (const start of [0, 25]) {
      try {
        const html = await fetchSearchPage(keywords, start);
        if (!html) continue;
        for (const item of parseListHtml(html)) {
          if (!withinLastWeek(item.postedAt)) continue;
          if (
            !isIsraelLocation(item.location, item.title, item.company) &&
            !/israel|ישראל/i.test(`${item.location} ${item.title}`)
          ) {
            continue;
          }
          const hay = `${item.title} ${item.company} ${item.location}`;
          if (!RELEVANT.test(hay)) continue;

          byId.set(item.id, {
            source: "linkedin",
            external_id: item.id,
            title: item.title,
            company: item.company || null,
            location: item.location || "Israel",
            url: item.url,
            description: `LinkedIn · ${item.title} @ ${item.company || "—"} · ${item.location || "Israel"}`,
            apply_email: null,
            post_kind: "job",
            channel: "linkedin",
            is_social: false,
            scraped_at: now,
            posted_at: item.postedAt
              ? new Date(item.postedAt).toISOString()
              : now,
          });
          if (byId.size >= maxJobs) break;
        }
      } catch {
        // network optional
      }
      if (byId.size >= maxJobs) break;
    }
    if (byId.size >= maxJobs) break;
  }

  const list = [...byId.values()];
  // Enrich a subset with full description for better matching + email extraction
  for (const job of list.slice(0, enrichN)) {
    const desc = await fetchJobDescription(job.external_id);
    if (desc.length > 80) {
      job.description = desc;
      job.apply_email = extractEmails(desc)[0] || null;
      const applyPage = extractExternalApplyUrl(desc);
      // Keep a working LinkedIn view URL for "open job"; stash ATS link in description
      // for web-apply (resolveApplyPageUrl reads description).
      const view = linkedInJobViewUrl(job.external_id);
      if (view) job.url = view;
      if (applyPage) {
        job.description = `Apply: ${applyPage}\n\n${desc}`;
      }
    } else if (isRealLinkedInJobId(job.external_id)) {
      const view = linkedInJobViewUrl(job.external_id);
      if (view) job.url = view;
    }
  }

  return list;
}

/** Delete LinkedIn rows older than maxAgeDays (default 7). */
export async function pruneOldLinkedInJobs(
  supabase: { from: (t: string) => any },
  maxAgeDays = 7,
): Promise<number> {
  const cutoff = new Date(
    Date.now() - maxAgeDays * 24 * 60 * 60 * 1000,
  ).toISOString();
  try {
    const { data } = await supabase
      .from("jobs")
      .select("id, posted_at, scraped_at")
      .eq("source", "linkedin")
      .limit(500);
    const stale = ((data || []) as Array<{
      id: string;
      posted_at?: string | null;
      scraped_at?: string | null;
    }>).filter((j) => {
      const ts = j.posted_at || j.scraped_at;
      return ts && ts < cutoff;
    });
    const ids = stale.map((j) => j.id);
    if (!ids.length) return 0;
    await supabase.from("jobs").delete().in("id", ids);
    return ids.length;
  } catch {
    return 0;
  }
}

File: job-agent/linkedin_jobs.py
"""LinkedIn active jobs — Israel, past 7 days (guest API)."""

from __future__ import annotations

import os
import re
from datetime import datetime, timedelta, timezone
from html import unescape
from typing import Any
from urllib.parse import urlencode

import httpx

from refresh import ScrapedJob

USER_AGENT = "Mozilla/5.0 (compatible; ai-agent-job-scanner/1.0)"
ISRAEL_GEO = "101620260"
PAST_WEEK = "r604800"

SEARCHES = [
    'AI OR LLM OR "machine learning" OR "data scientist" OR "בינה מלאכותית"',
    '"AI product manager" OR "AI product owner" OR "product manager AI" OR "LLM product" OR "מנהל מוצר AI"',
    "אנליסט OR חשב OR \"הצלחת לקוחות\" OR שיווק OR \"product analyst\"",
    'finance OR fintech OR FP&A OR controller OR "financial analyst" OR פיננסים',
    '"product owner" OR "product marketing" OR "customer success" OR operations',
]

RELEVANT = re.compile(
    r"ai|llm|machine learning|data scientist|data analyst|product|finance|fintech|"
    r"fp&a|controller|analyst|marketing|sales|customer success|"
    r"מוצר|פיננס|בינה|שיווק|מכירות|אנליסט",
    re.I,
)

EMAIL_RE = re.compile(r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}")


def _strip(html: str) -> str:
    text = re.sub(r"<[^>]+>", " ", html)
    text = unescape(text)
    return re.sub(r"\s+", " ", text).strip()


def _parse_list(html: str) -> list[dict[str, Any]]:
    jobs: list[dict[str, Any]] = []
    parts = re.split(r'data-entity-urn="urn:li:jobPosting:', html)[1:]
    for part in parts:
        m_id = re.match(r"(\d+)", part)
        if not m_id:
            continue
        job_id = m_id.group(1)
        title_m = re.search(r"base-search-card__title[^>]*>([\s\S]*?)</h3>", part, re.I)
        company_m = re.search(
            r"base-search-card__subtitle[\s\S]*?<a[^>]*>([\s\S]*?)</a>",
            part,
            re.I,
        )
        loc_m = re.search(r"job-search-card__location[^>]*>([\s\S]*?)</span>", part, re.I)
        href_m = re.search(r'base-card__full-link[^>]*href="([^"]+)"', part, re.I)
        time_m = re.search(r'<time[^>]*datetime="([^"]+)"', part, re.I)
        title = _strip(title_m.group(1)) if title_m else ""
        if not title:
            continue
        href = unescape(href_m.group(1)) if href_m else f"https://www.linkedin.com/jobs/view/{job_id}"
        url = href.split("?")[0]
        jobs.append(
            {
                "id": job_id,
                "title": title,
                "company": _strip(company_m.group(1)) if company_m else "",
                "location": _strip(loc_m.group(1)) if loc_m else "Israel",
                "url": url,
                "posted_at": time_m.group(1) if time_m else None,
            }
        )
    return jobs


def _within_week(posted_at: str | None) -> bool:
    if not posted_at:
        return True
    try:
        dt = datetime.fromisoformat(posted_at.replace("Z", "+00:00"))
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return datetime.now(timezone.utc) - dt <= timedelta(days=7, hours=12)
    except ValueError:
        return True


def _fetch_description(client: httpx.Client, job_id: str) -> str:
    try:
        resp = client.get(
            f"https://www.linkedin.com/jobs-guest/jobs/api/jobPosting/{job_id}",
            timeout=8.0,
        )
        if resp.status_code != 200:
            return ""
        html = resp.text
        m = re.search(r"show-more-less-html__markup[\s\S]*?>([\s\S]*?)</div>", html, re.I)
        return _strip(m.group(1) if m else html)[:4000]
    except Exception:  # noqa: BLE001
        return ""


def fetch_linkedin_israel_jobs(max_jobs: int = 80, enrich: int = 20) -> list[ScrapedJob]:
    now = datetime.now(timezone.utc)
    by_id: dict[str, ScrapedJob] = {}
    headers = {"User-Agent": USER_AGENT, "Accept": "text/html"}

    try:
        with httpx.Client(timeout=25.0, headers=headers, follow_redirects=True) as client:
            for keywords in SEARCHES:
                for start in (0, 25):
                    params = {
                        "keywords": keywords,
                        "location": "Israel",
                        "geoId": ISRAEL_GEO,
                        "f_TPR": PAST_WEEK,
                        "start": str(start),
                        "count": "25",
                    }
                    url = (
                        "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search?"
                        + urlencode(params)
                    )
                    try:
                        resp = client.get(url)
                        if resp.status_code != 200:
                            continue
                        for item in _parse_list(resp.text):
                            if not _within_week(item.get("posted_at")):
                                continue
                            hay = f"{item['title']} {item['company']} {item['location']}"
                            if not RELEVANT.search(hay):
                                continue
                            if not re.search(r"israel|ישראל|tel aviv|herzliya|haifa|jerusalem", hay, re.I):
                                # guest search is geo-scoped; still require IL signal
                                if "Israel" not in (item.get("location") or ""):
                                    continue
                            posted = None
                            if item.get("posted_at"):
                                try:
                                    posted = datetime.fromisoformat(
                                        item["posted_at"].replace("Z", "+00:00")
                                    )
                                    if posted.tzinfo is None:
                                        posted = posted.replace(tzinfo=timezone.utc)
                                except ValueError:
                                    posted = now
                            else:
                                posted = now

                            by_id[item["id"]] = ScrapedJob(
                                source="linkedin",
                                external_id=item["id"],
                                title=item["title"],
                                company=item["company"] or None,
                                location=item["location"] or "Israel",
                                url=item["url"],
                                description=(
                                    f"LinkedIn · {item['title']} @ {item['company'] or '—'} · "
                                    f"{item['location'] or 'Israel'}"
                                ),
                                posted_at=posted,
                                apply_email=None,
                                post_kind="job",
                                channel="linkedin",
                                is_social=False,
                            )
                            if len(by_id) >= max_jobs:
                                break
                    except Exception:  # noqa: BLE001
                        continue
                    if len(by_id) >= max_jobs:
                        break
                if len(by_id) >= max_jobs:
                    break

            # Enrich descriptions for matching quality
            for job in list(by_id.values())[:enrich]:
                desc = _fetch_description(client, job.external_id)
                if len(desc) > 80:
                    job.description = desc
                    emails = EMAIL_RE.findall(desc)
                    if emails:
                        # Strip RTL/ZW marks that break Resend `to` validation
                        raw = re.sub(
                            r"[\u200B-\u200F\u202A-\u202E\u2060\u2066-\u2069\uFEFF]",
                            "",
                            emails[0],
                        )
                        job.apply_email = raw.lower()
    except Exception:  # noqa: BLE001
        return list(by_id.values())

    return list(by_id.values())


def prune_old_linkedin_jobs(client: Any, max_age_days: int = 7) -> int:
    cutoff = datetime.now(timezone.utc) - timedelta(days=max_age_days)
    try:
        rows = (
            client.table("jobs")
            .select("id, posted_at, scraped_at")
            .eq("source", "linkedin")
            .limit(500)
            .execute()
            .data
            or []
        )
    except Exception:  # noqa: BLE001
        return 0
    stale: list[str] = []
    for row in rows:
        ts = row.get("posted_at") or row.get("scraped_at")
        if not ts:
            continue
        try:
            dt = datetime.fromisoformat(str(ts).replace("Z", "+00:00"))
            if dt.tzinfo is None:
                dt = dt.replace(tzinfo=timezone.utc)
            if dt < cutoff:
                stale.append(row["id"])
        except ValueError:
            continue
    if not stale:
        return 0
    try:
        client.table("jobs").delete().in_("id", stale).execute()
    except Exception:  # noqa: BLE001
        return 0
    return len(stale)
